### ISSUE TITLE
Some allocation sizes fail

### ISSUE THREAD

**Issue by delanyinspiron6400** [canonizer/halloc#3]
Currently trying to test halloc on the TITAN V with CUDA 10.2 on Linux, compiling once with and without ITS (Independent Thread Scheduling).
I try to run two sweeping test cases, once with 10K and once with 100K threads trying to allocate memory, for 10K only allocations between 16B - 256B work, larger than that (except where the delegation to CUDA happens) and the code hangs, for 100K allocations only 16 & 32B work, all larger than that once again hang.
I also tried instead of uniformly allocating the same size for each thread, alternating the allocation size between even and odd threads between 16/32 Bytes and this also does not seem to work, although purely 16B or 32B allocations do work for limited amounts of threads!

any clues what might be going on here?

**Comment by LAhmos** [comment]
I have the same issue, did you figure it out?

**Comment by delanyinspiron6400** [comment]
Yes, the problem are the internal synchronization primitives, which presuppose warp-synchronous execution. 
Since they are buried quite deep in the logic, the simplest solution is to force warp-synchronous behavior, by compiling using `-gencode=arch=compute_60,code=sm_70` (if you want to compile for the Volta architecture but with no ITS, simply replace the `sm_70` with your newer architecture).
Unfortunately, this probably means that, without update, this will stop working soon, as CUDA will not compile with the old *non-sync* warp-level primitives in place.
But for now, that's how it works.

**Comment by LAhmos** [comment]
I tried -gencode=arch=compute_60,code=sm_70 , but it does not seem to work. what other flags are you passing to nvcc ?

**Comment by delanyinspiron6400** [comment]
Nothing that should be affecting this issue, unfortunately. 
Another thing you can try is to just allocate with one thread per warp, which beforehand compute the sum of all allocation requests within the warp.
But depending on your use case, this can make memory management quite cumbersome if you want to free individual allocations, which are then part of larger allocations.
